src/test/fuzz/util/net.cpp

```diff
@@ -25,29 +25,57 @@
+CNetAddr ConsumeNetAddr(FuzzedDataProvider& fuzzed_data_provider, FastRandomContext* rand) noexcept
 {
-    const Network network = fuzzed_data_provider.PickValueInArray({Network::NET_IPV4, Network::NET_IPV6, Network::NET_INTERNAL, Network::NET_ONION});
-    CNetAddr net_addr;
-    if (network == Network::NET_IPV4) {
-        in_addr v4_addr = {};
-        v4_addr.s_addr = fuzzed_data_provider.ConsumeIntegral<uint32_t>();
-        net_addr = CNetAddr{v4_addr};
-    } else if (network == Network::NET_IPV6) {
-        if (fuzzed_data_provider.remaining_bytes() >= 16) {
-            in6_addr v6_addr = {};
-            memcpy(v6_addr.s6_addr, fuzzed_data_provider.ConsumeBytes<uint8_t>(16).data(), 16);
-            net_addr = CNetAddr{v6_addr, fuzzed_data_provider.ConsumeIntegral<uint32_t>()};
+    struct NetAux {
+        Network net;
+        CNetAddr::BIP155Network bip155;
+        size_t len;
+    };
+
+    static constexpr std::array<NetAux, 6> nets{
+        NetAux{.net = Network::NET_IPV4, .bip155 = CNetAddr::BIP155Network::IPV4, .len = ADDR_IPV4_SIZE},
+        NetAux{.net = Network::NET_IPV6, .bip155 = CNetAddr::BIP155Network::IPV6, .len = ADDR_IPV6_SIZE},
+        NetAux{.net = Network::NET_ONION, .bip155 = CNetAddr::BIP155Network::TORV3, .len = ADDR_TORV3_SIZE},
+        NetAux{.net = Network::NET_I2P, .bip155 = CNetAddr::BIP155Network::I2P, .len = ADDR_I2P_SIZE},
+        NetAux{.net = Network::NET_CJDNS, .bip155 = CNetAddr::BIP155Network::CJDNS, .len = ADDR_CJDNS_SIZE},
+        NetAux{.net = Network::NET_INTERNAL, .bip155 = CNetAddr::BIP155Network{0}, .len = 0},
+    };
+
+    const size_t i{rand == nullptr
+        ? fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, nets.size() - 1)
+        : static_cast<size_t>(rand->randrange(nets.size()))};
+
+    const auto& aux = nets[i];
+
+    CNetAddr addr;
+
+    if (aux.net == Network::NET_INTERNAL) {
+        if (rand == nullptr) {
+            addr.SetInternal(fuzzed_data_provider.ConsumeBytesAsString(32));
+        } else {
+            const auto v = rand->randbytes(32);
+            addr.SetInternal(std::string{v.begin(), v.end()});
         }
-    } else if (network == Network::NET_INTERNAL) {
-        net_addr.SetInternal(fuzzed_data_provider.ConsumeBytesAsString(32));
-    } else if (network == Network::NET_ONION) {
-        auto pub_key{fuzzed_data_provider.ConsumeBytes<uint8_t>(ADDR_TORV3_SIZE)};
-        pub_key.resize(ADDR_TORV3_SIZE);
-        const bool ok{net_addr.SetSpecial(OnionToString(pub_key))};
-        assert(ok);
+        return addr;
     }
-    return net_addr;
+
+    CDataStream s(SER_NETWORK, PROTOCOL_VERSION | ADDRV2_FORMAT);
+
+    s << static_cast<uint8_t>(aux.bip155);
+
+    std::vector<uint8_t> data;
+    if (rand == nullptr) {
+        data = fuzzed_data_provider.ConsumeBytes<uint8_t>(aux.len);
+        data.resize(aux.len);
+    } else {
+        data = rand->randbytes(aux.len);
+    }
+    s << data;
+
+    s >> addr;
```

Comment:
I tried to understand the CI fail: 
The process of unserializing can change the network into `NET_UNROUTABLE`, for example if a CJDNS address with the wrong prefix was generated. As a result of that, the `fuzz/banman` target can hit https://github.com/bitcoin/bitcoin/blob/7753efcbcf3ca2e1b46f167936e11cc580e28c7a/src/test/fuzz/banman.cpp#L112
because while converting the unroutable address to a `CSubNet`, no netmask is added as for [IPv6](https://github.com/bitcoin/bitcoin/blob/7753efcbcf3ca2e1b46f167936e11cc580e28c7a/src/netaddress.cpp#L1023), but after serializing / unserializing via banlist, it's no longer seen as unroutable but IPv6 and gets a /128 netmask.

Reply:
Slightly unrelated, but while looking into this I got the impression that banning CJDNS peer doesn't work at all currently (no `MaybeFlipIPv6toCJDNS`). I might try to add this possibility in another PR.

Reply:
Alright, this PR uncovered a problem elsewhere!

You are right in your analysis (except that `NET_UNROUTABLE` is not involved): the test bans `CNetAddr{.m_net = NET_CJDNS, m_addr = 91b4:919d...}` (that is `!IsValid()` because it does not start with `fc`). Then it gets serialized in `fuzzed_banlist.json` like:
```json
        {   
            "version": 1,
            "ban_created": 2984688890,
            "banned_until": 3982933363,
            "address": "91b4:919d:1191:9140:0:c4:207f:7f00"
        }   
```
and then of course, after unserialize ends up as an IPv6 address in `banmap_read`. This is something I have overlooked wrt `CJDNS` and `banlist.json`.

So, two issues:

1. How to handle an attempt to ban an address that is `!IsValid()`?
    * Refuse and change `BanMan::Ban()` to return `bool` / failure, or
    * Don't change anything, leave it as it is now.

2. How to survive CJDNS addresses through banman ser/unser aka `banlist.json`?
    * Save the network explicitly in a new field in the JSON, or
    * During unserialize maybe flip an IPv6 address that starts with `fc` to CJDNS type if `-cjdnsreachable` (but this would mean the contents of `banlist.json` could be interpreted differently based on the environment where it is loaded).

Reply:
I think that the fuzz test will not only fail on invalid addresses, it will also fail on valid CJDNS addresses for the same reason. The important thing is that IPv6 addresses will get the `/128` netmask, other networks won't, and if the address, valid or not, is re-interpreted as IPv6 after unserialization of `banlist.json`, the assert will fail.

> During unserialize maybe flip an IPv6 address that starts with fc to CJDNS type if -cjdnsreachable (but this would mean the contents of banlist.json could be interpreted differently based on the environment where it is loaded).

I had this in mind, and I verified that banning CJDNS peers will work that way. The different interpretation may not a huge problem, because regular IPv6 peers can't be from that reserved range, so there would be no risk of banning the wrong peer after disabling `-cjdnsreachable`. On the other hand, we would need some logic to deal with the fact that when interpreted as `IPv6`, the address gets a `/128` mask and switching back to `-cjdnsreachable` won't work anymore, so maybe the cleanest solution would be to save the network explicitly?

Reply:
This turned out to be somewhat involved change. Opened a PR at https://github.com/bitcoin/bitcoin/pull/27071.

Reply:
Resolving this, given that https://github.com/bitcoin/bitcoin/pull/27071 has now been merged.